### ISSUE TITLE
Is this a leak or something?

### ISSUE THREAD

**Issue by zxqfox** [reactphp/http#27]
Versions:
```
$ uname -a
Linux dev.qfox.ru 3.2.0-4-486 #1 Debian 3.2.51-1 i686 GNU/Linux
$ php -v
PHP 5.4.38-1~dotdeb.1 (cli) (built: Feb 19 2015 22:21:23)
Copyright (c) 1997-2014 The PHP Group
Zend Engine v2.4.0, Copyright (c) 1998-2014 Zend Technologies
    with Xdebug v2.3.1, Copyright (c) 2002-2015, by Derick Rethans
```

I have a simple file app.php:
```php
require 'vendor/autoload.php';

$app = function ($request, $response) {
    $response->writeHead(200, array('Content-Type' => 'text/plain'));
    $response->end("Hello World\n");
};

$loop = React\EventLoop\Factory::create();
$socket = new React\Socket\Server($loop);
$http = new React\Http\Server($socket, $loop);

$http->on('request', $app);
echo "Server running at http://0.0.0.0:1337\n";

$socket->listen(1337, '0.0.0.0');
$loop->run();
```

With installed `react/http` (here is my composer.json):
```
{
    "require": {
        "react/http": "0.3.*"
    }
}
```

Running it:
```
$ php app.php
Server running at http://127.0.0.1:1337
```

And sending 10k requests:
```
$ ab -n 10000 -c 500 http://localhost:1337/
This is ApacheBench, Version 2.3 <$Revision: 655654 $>
Copyright 1996 Adam Twiss, Zeus Technology Ltd, http://www.zeustech.net/
Licensed to The Apache Software Foundation, http://www.apache.org/

Benchmarking localhost (be patient)
Completed 1000 requests
Completed 2000 requests
Completed 3000 requests
Completed 4000 requests
Completed 5000 requests
Completed 6000 requests
Completed 7000 requests
Completed 8000 requests
Completed 9000 requests
apr_socket_recv: Connection reset by peer (104)
Total of 9953 requests completed
```

Why some of requests was reseted by peer? Something is wrong?

**Comment by tomplus** [comment]
I have the same problem. The program suddenly exits and prints the exception

     {"error":{"code":-100,"data":{"message":"Response head has already been written."}}}

I use the latest version of packages:

  - Installing react/promise (v2.2.1)
  - Installing react/cache (v0.4.0)
  - Installing react/stream (v0.4.2)
  - Installing react/event-loop (v0.4.1)
  - Installing react/socket (v0.4.2)
  - Installing react/http (v0.4.1)
  - Installing react/dns (v0.4.1)
  - Installing react/socket-client (v0.4.3)
  - Installing react/http-client (v0.4.4)

Any ideas ?

**Comment by skedone** [comment]
@zxqfox as far as I know, your error means you hit the socket connections limit of your server.
It should not be an issue for reactphp/http but you need to tune your server (especially sysctl - https://rtcamp.com/tutorials/linux/sysctl-conf/)

**Comment by zxqfox** [comment]
@skedone Just there are no troubles with nodejs, or nginx, so I'm gonna think that ReactPHP closing connections much slower, or something like that.

Btw `-c 500` means just 500 concurrent connections. 10k is 20x more than that limit. Are you sure I should tune my server to get 500 connections with ReactPHP?

**Comment by tomplus** [comment]
I've caught this problem with -c 10 so it's not a problem with sysctl limitation.

**Comment by skedone** [comment]
@zxqfox mmm I tested the same script against this configuration and it passed on a base DigitalOcean's VPS but it has a compiled and tuned version for PHP (production env).

Then I tested on my laptop and I encountered you exact error.
After this I tested the library against react/socket (without the http module) and I got the same error.

I think this issue should be move to reactphp/socket because it doesn't belongs to the http wrapper but to the socket itself.

@tomplus are you using the same snippet of @zxqfox or another one?
